[SVG2] Remove non-standard `animateColor`

Ahmad-S792 opened the pull request:
#### 1674e8f0e5b811710ca6f0b03e4d708a53768958
<pre>
[SVG2] Remove non-standard `animateColor`

<a href="https://bugs.webkit.org/show_bug.cgi?id=191370">https://bugs.webkit.org/show_bug.cgi?id=191370</a>
<a href="https://rdar.apple.com/problem/122586568">rdar://problem/122586568</a>

Reviewed by Tim Nguyen and Anne van Kesteren.

This patch is to align WebKit with Gecko / Firefox, Blink / Chromium
and Web Specification [1]:

[1] <a href="https://svgwg.org/specs/animations/#ChangesSinceSVG11">https://svgwg.org/specs/animations/#ChangesSinceSVG11</a>

&quot;Removed the animateColor element.&quot;

It was removed by Blink in 2014 while Firefox never supported this,
since all use cases can be tackled with `animate`.

&gt; Compiler &amp; Build System Changes:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:

&gt; Deleted Files:
* Source/WebCore/svg/SVGAnimateColorElement.cpp:
* Source/WebCore/svg/SVGAnimateColorElement.h:
* Source/WebCore/svg/SVGAnimateColorElement.idl:

&gt; Modified:
* Source/WebCore/svg/SVGAnimateElementBase.cpp:
(SVGAnimateElementBase::SVGAnimateElementBase):
* Source/WebCore/svg/SVGAnimateElementBase.h:
(isType):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(SVGAnimationElement::currentValuesForValuesAnimation):
* Source/WebCore/svg/svgtags.in: Retain `parsing rules`

&gt; Tests &amp; Test Expectations Update:
* LayoutTests/imported/w3c/web-platform-tests/svg/import/animate-elem-23-t-manual.svg:
* LayoutTests/imported/w3c/web-platform-tests/svg/import/animate-elem-84-t-manual.svg:
* LayoutTests/svg/animations/animateColor-additive-2a-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2b-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2c-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2d-expected.svg:
* LayoutTests/svg/animations/resources/additive-type-by-animation.svg:
* LayoutTests/svg/custom/animation-currentColor.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-23-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-84-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-85-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/pservers-grad-19-b.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-30-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-85-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-pservers-grad-01-b-manual-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-30-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-85-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-pservers-grad-01-b-manual-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/svg/animations/animate-montion-invalid-attribute.svg:
* LayoutTests/svg/custom/global-constructors-expected.txt:
* LayoutTests/svg/custom/global-constructors.html:
* LayoutTests/svg/dom/svg2-inheritance-expected.txt:
* LayoutTests/svg/dom/svg2-inheritance.html:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt: Added new expectation

Canonical link: <a href="https://commits.webkit.org/276683@main">https://commits.webkit.org/276683@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65ab250773662b4e6b588d61c31a02b85b7512b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47734 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41083 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28280 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21584 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36977 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21247 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18077 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39951 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3122 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40254 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49419 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20048 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16575 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44067 "Found 45 new test failures: fast/dom/Window/window-lookup-precedence.html, fast/dom/dom-constructors.html, fast/events/event-handler-detached-document.html, http/tests/xmlhttprequest/post-content-type-document.html, http/tests/xmlhttprequest/xhr-and-parse-large-document.html, imported/w3c/web-platform-tests/custom-elements/parser/parser-uses-registry-of-owner-document.html, imported/w3c/web-platform-tests/fetch/private-network-access/nested-worker.https.window.html, imported/w3c/web-platform-tests/fetch/private-network-access/nested-worker.window.html, imported/w3c/web-platform-tests/html/dom/idlharness.worker.html, imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/nested-worker-success.https.html ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42856 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10081 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21710 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21042 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/svg/svgtags.in:
```diff
@@ -8,7 +8,7 @@ altGlyph JSInterfaceName=SVGElement
 altGlyphItem JSInterfaceName=SVGElement
 animate
-animateColor
+animateColor interfaceName=SVGUnknownElement, JSInterfaceName=SVGElement, noConstructor
```
Can we remove this?

Ahmad-S792 replied:
We can't since HTML Parsing rules needed to kept intact. I removed in first PR and it failed bunch of WPT tests in relation to HTML5 Parser.

nt1m replied:
@annevk Is this true?

Ahmad-S792 replied:
Just for reference - https://ews-build.webkit.org/#/builders/1/builds/42456 (If we look into failures here).

nt1m replied:
So the only thing that changes if we remove this is the casing which goes from `animateColor` to `animatecolor`. Seems removable to me, maybe @annevk has guidance on what to do with the tests.

annevk replied:
There's two things here:

1. We should move the case-changing logic into the HTML parser instead. I already did this for attributes, but I couldn't figure out how to do elements properly: https://bugs.webkit.org/show_bug.cgi?id=254881.
2. We could possibly look into changing the HTML Standard to no longer recognize `animateColor` as an element, but that kind of parser churn is best avoided I think. Especially as there's no real good reason for it. It's just that in WebKit we have this unfortunate architecture where the case-changing depends on lists maintained outside of the HTML parser.

annevk replied:
I do wonder if we need to set all these interface name thingies. Does anything break if you just leave it as "animateColor"?

Ahmad-S792 replied:
@annevk - It leads to build error:
```
/Users/ahmadsaleem/Documents/GitHub-Webkit-Ahmad-Fork/Untitled/WebKitBuild/Release/DerivedSources/WebCore/SVGElementFactory.cpp:40:10: fatal error:
      'SVGAnimateColorElement.h' file not found
#include "SVGAnimateColorElement.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Ahmad-S792 replied:
Since it leads to compiler issue, I am considering this as 'RESOLVED' unless if in future, we update HTML5 Parser tests and we can remove it completely.

nt1m commented on LayoutTests/svg/dom/svg2-inheritance-expected.txt:
```diff
@@ -5,7 +5,7 @@ On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE
 PASS SVGAElement inherits SVGGraphicsElement
 PASS SVGAngle inherits Object
-PASS SVGAnimateColorElement inherits SVGAnimationElement
+FAIL SVGAnimateColorElement is not defined
```
Can you change the test so this doesn't appear as a failure? (either remove the subtest or change to test that it's not defined)

Ahmad-S792 replied:
I have separate bug already to fix this because we have multiple such cases in this tests and @rwlbuis raised also good question whether we have WPT test for this or we can make one. So I thought to tackle it separately.

annevk replied:
I don't think we should tackle that completely separately. We want to be able to revert without much hassle.

Ahmad-S792 replied:
Removed.

nt1m commented on LayoutTests/svg/custom/global-constructors-expected.txt:
```diff
@@ -175,7 +175,7 @@ PASS SVGAnimateElement.toString() is "function SVGAnimateElement() {\n    [nativ
 PASS SVGAnimateMotionElement.toString() is "function SVGAnimateMotionElement() {\n    [native code]\n}"
 PASS SVGMPathElement.toString() is "function SVGMPathElement() {\n    [native code]\n}"
-PASS SVGAnimateColorElement.toString() is "function SVGAnimateColorElement() {\n    [native code]\n}"
+PASS SVGAnimateColorElement threw exception ReferenceError: Can't find variable: SVGAnimateColorElement.
```
Can you change the subtest to test this is undefined or remove it?

Ahmad-S792 replied:
Can remove it. 👍

Ahmad-S792 replied:
Removed.

annevk reviewed (approved):
I don't really like the svgtags situation, but I don't want to block this on that.

Commit:
[SVG2] Remove non-standard `animateColor`

https://bugs.webkit.org/show_bug.cgi?id=191370
rdar://problem/122586568

Reviewed by Tim Nguyen and Anne van Kesteren.

This patch is to align WebKit with Gecko / Firefox, Blink / Chromium
and Web Specification [1]:

[1] https://svgwg.org/specs/animations/#ChangesSinceSVG11

"Removed the animateColor element."

It was removed by Blink in 2014 while Firefox never supported this,
since all use cases can be tackled with `animate`.

> Compiler & Build System Changes:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:

> Deleted Files:
* Source/WebCore/svg/SVGAnimateColorElement.cpp:
* Source/WebCore/svg/SVGAnimateColorElement.h:
* Source/WebCore/svg/SVGAnimateColorElement.idl:

> Modified:
* Source/WebCore/svg/SVGAnimateElementBase.cpp:
(SVGAnimateElementBase::SVGAnimateElementBase):
* Source/WebCore/svg/SVGAnimateElementBase.h:
(isType):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(SVGAnimationElement::currentValuesForValuesAnimation):
* Source/WebCore/svg/svgtags.in: Retain `parsing rules`

> Tests & Test Expectations Update:
* LayoutTests/imported/w3c/web-platform-tests/svg/import/animate-elem-23-t-manual.svg:
* LayoutTests/imported/w3c/web-platform-tests/svg/import/animate-elem-84-t-manual.svg:
* LayoutTests/svg/animations/animateColor-additive-2a-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2b-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2c-expected.svg:
* LayoutTests/svg/animations/animateColor-additive-2d-expected.svg:
* LayoutTests/svg/animations/resources/additive-type-by-animation.svg:
* LayoutTests/svg/custom/animation-currentColor.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-23-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-84-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/animate-elem-85-t.svg:
* LayoutTests/svg/W3C-SVG-1.1/pservers-grad-19-b.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-30-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-85-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-pservers-grad-01-b-manual-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-30-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-85-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-pservers-grad-01-b-manual-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt:
* LayoutTests/svg/animations/animate-montion-invalid-attribute.svg:
* LayoutTests/svg/custom/global-constructors-expected.txt:
* LayoutTests/svg/custom/global-constructors.html:
* LayoutTests/svg/dom/svg2-inheritance-expected.txt:
* LayoutTests/svg/dom/svg2-inheritance.html:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/animate-elem-85-t-expected.txt: Added new expectation

Canonical link: https://commits.webkit.org/276683@main
